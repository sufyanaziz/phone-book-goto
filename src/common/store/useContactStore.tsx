import { createContext, useState } from "react";

type TContactStore = {
  contact: { [x: string]: unknown };
  search: string;
  offset: number;
  modal: {
    name: string;
    data: { [key: string]: any };
  };
  onSetContact: (contact: { [x: string]: unknown }) => void;
  onSetSearch: (search: string) => void;
  onSetOffset: (offset: number) => void;
  closeModal: () => void;
  openModal: (modalName: string, modalData: any) => void;
};

const ContactStore = createContext<TContactStore>({
  contact: {},
  search: "",
  offset: 0,
  modal: {
    name: "",
    data: {},
  },
  onSetContact: () => {},
  onSetSearch: () => {},
  onSetOffset: () => {},
  closeModal: () => {},
  openModal: () => {},
});

type TContactProvider = {
  children?: React.ReactNode;
};

const ContactProvider = ({ children }: TContactProvider) => {
  const [contact, setContact] = useState({});
  const [search, setSearch] = useState<string>("");
  const [offset, setOffset] = useState<number>(0);
  const [modal, setModal] = useState({ name: "", data: {} });
  return (
    <ContactStore.Provider
      value={{
        contact,
        onSetContact: setContact,
        search,
        onSetSearch: (search) => {
          setSearch(search);
          setOffset(0);
        },
        offset,
        onSetOffset: setOffset,
        modal,
        openModal: (modalName, modalData) => {
          setModal({ name: modalName, data: modalData });
        },
        closeModal: () => {
          setModal({ name: "", data: {} });
        },
      }}
    >
      {children}
    </ContactStore.Provider>
  );
};

const ContactConsumer = ContactStore.Consumer;

export { ContactStore, ContactProvider, ContactConsumer };
