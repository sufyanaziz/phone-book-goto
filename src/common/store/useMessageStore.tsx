import { TMessageId } from "@common/types/message";
import { createContext, useState } from "react";

type TMessageStore = {
  showMessage: boolean;
  activeId: TMessageId;
  setShowMessage: (activeId: TMessageId, showMessage: boolean) => void;
};

const MessageStore = createContext<TMessageStore>({
  showMessage: false,
  activeId: "",
  setShowMessage: () => {},
});

type TMessageProvider = {
  children?: React.ReactNode;
};

const MessageProvider = ({ children }: TMessageProvider) => {
  const [message, setMessage] = useState<{
    showMessage: boolean;
    activeId: TMessageId;
  }>({
    showMessage: false,
    activeId: "",
  });

  return (
    <MessageStore.Provider
      value={{
        showMessage: message.showMessage,
        activeId: message.activeId,
        setShowMessage: (activeId, showMessage) => {
          setMessage({
            activeId,
            showMessage,
          });
        },
      }}
    >
      {children}
    </MessageStore.Provider>
  );
};

const ContactConsumer = MessageStore.Consumer;

export { MessageStore, MessageProvider, ContactConsumer };
