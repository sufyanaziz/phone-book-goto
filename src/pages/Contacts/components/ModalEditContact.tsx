import React, { useContext, useEffect, useMemo, useState } from "react";
import Modal from "@common/components/Modal";
import { ContactStore } from "@common/store/useContactStore";
import { MODAL_NAME } from "../constant/modal";
import Text from "@common/components/Text";
import { css } from "@emotion/react";
import Input from "@common/components/Input";
import Button from "@common/components/Button";

type TModalEditContact = {
  onEditContact: (id: number, firstName: string, lastName: string) => void;
  loading: boolean;
  onClose: () => void;
  isComplete: boolean;
};

const ModalEditContact = ({
  onEditContact,
  loading,
  onClose,
  isComplete,
}: TModalEditContact) => {
  const [firstName, setFirstName] = useState("");
  const [lastName, setLastName] = useState("");
  const [phoneNumbers, setPhoneNumbers] = useState<string[]>([""]);

  const { modal, closeModal } = useContext(ContactStore);

  const isDisabled = useMemo(() => {
    if (loading) return true;
    if (!firstName) return true;
    if (!lastName) return true;
    if (phoneNumbers.some((phone) => !phone)) return true;
    return false;
  }, [firstName, lastName, phoneNumbers, loading]);

  const handleCloseModal = () => {
    onClose();
    closeModal();
    setFirstName("");
    setLastName("");
    setPhoneNumbers([""]);
  };

  useEffect(() => {
    if (modal.name === MODAL_NAME.modalEditContact) {
      setFirstName(modal.data.first_name);
      setLastName(modal.data.last_name);
      setPhoneNumbers(
        modal.data.phones?.map((phone: { number: string }) => phone.number)
      );
    }
  }, [modal.name]);

  useEffect(() => {
    if (!loading && isComplete) {
      handleCloseModal();
    }
  }, [isComplete, loading]);

  const modalEditContactStyle = () => {
    return css({
      padding: "12px 0px",
      ".section-name": {
        ".first-name, .last-name": {
          marginBottom: "8px",
          width: "100%",
          ".input": {
            width: "100%",
          },
        },
      },
      ".header-title": {
        marginBottom: "4px",
      },
      ".section-phone-number": {
        ".phone-container": {
          display: "flex",
          marginBottom: "8px",
          width: "100%",
          "&:last-child": {
            marginBottom: "0px",
          },
          ".icons": {
            display: "flex",
            marginLeft: "8px",
            ".add-icon, .minus-icon": {
              display: "flex",
              alignItems: "center",
            },
          },
        },
      },
    });
  };

  return (
    <Modal
      isVisible={modal.name === MODAL_NAME.modalEditContact}
      header="Edit Contact"
      onClose={handleCloseModal}
      content={
        <div css={modalEditContactStyle}>
          <div className="section-name">
            <div className="first-name">
              <Text className="header-title" text="First Name" type="medium" />
              <Input
                className="input"
                placeholder="Input First Name..."
                acceptValue="string-number"
                value={firstName}
                onChange={setFirstName}
              />
            </div>
            <div className="last-name">
              <Text className="header-title" text="Last Name" type="medium" />
              <Input
                className="input"
                placeholder="Input Last Name..."
                acceptValue="string-number"
                value={lastName}
                onChange={setLastName}
              />
            </div>
          </div>
          <div className="section-phone-number">
            <Text className="header-title" text="Phones" type="medium" />
            {phoneNumbers.map((phone, listIdx) => {
              return (
                <div className="phone-container" key={`phone-${listIdx}`}>
                  <Input
                    className="input"
                    placeholder={`Phone ${listIdx + 1}`}
                    value={phone}
                    acceptValue="number-only"
                    disabled
                    onChange={(value) => {
                      const phones = [...phoneNumbers];
                      phones[listIdx] = value;
                      setPhoneNumbers(phones);
                    }}
                  />
                </div>
              );
            })}
          </div>
        </div>
      }
      footer={
        <div className="flex" css={{ padding: "0 12px 12px 12px" }}>
          <Button
            label="Submit"
            onClick={() => onEditContact(modal.data.id, firstName, lastName)}
            disabled={isDisabled}
          />
        </div>
      }
    />
  );
};

export default ModalEditContact;
