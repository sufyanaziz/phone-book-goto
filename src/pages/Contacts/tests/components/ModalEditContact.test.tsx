import React from "react";
import renderer from "react-test-renderer";
import ModalEditContact from "@pages/Contacts/components/ModalEditContact";
import { ContactStore } from "@common/store/useContactStore";
import { mockStoreData } from "../mock/mockStore";

describe("Component ModalEditContact", () => {
  it("should render ModalEditContact component correctly", () => {
    const component = renderer.create(
      <ContactStore.Provider
        value={{
          ...mockStoreData,
          modal: {
            name: "modal-edit-contact",
            data: {
              id: 9,
              first_name: "first name",
              last_name: "dummy last name",
              phones: [{ number: "6785421" }],
            },
          },
        }}
      >
        <ModalEditContact
          isComplete
          loading
          onClose={jest.fn()}
          onEditContact={jest.fn()}
        />
      </ContactStore.Provider>
    );
    expect(component).toMatchSnapshot();
  });
});
