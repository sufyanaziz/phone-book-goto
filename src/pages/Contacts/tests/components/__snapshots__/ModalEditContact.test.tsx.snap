// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component ModalEditContact should render ModalEditContact component correctly 1`] = `
<div
  css={[Function]}
>
  <div
    className="modal-mask"
    onClick={[Function]}
  />
  <div
    className="modal-content"
  >
    <div
      className="header"
    >
      <div
        className=""
        css={[Function]}
      >
        Edit Contact
      </div>
      <div
        className="close-icon"
        onClick={[Function]}
      >
        <div
          className=""
          css={[Function]}
        >
          X
        </div>
      </div>
    </div>
    <div
      className="content"
    >
      <div
        css={[Function]}
      >
        <div
          className="section-name"
        >
          <div
            className="first-name"
          >
            <div
              className="header-title"
              css={[Function]}
            >
              First Name
            </div>
            <input
              className="input"
              disabled={false}
              onChange={[Function]}
              placeholder="Input First Name..."
              value=""
            />
          </div>
          <div
            className="last-name"
          >
            <div
              className="header-title"
              css={[Function]}
            >
              Last Name
            </div>
            <input
              className="input"
              disabled={false}
              onChange={[Function]}
              placeholder="Input Last Name..."
              value=""
            />
          </div>
        </div>
        <div
          className="section-phone-number"
        >
          <div
            className="header-title"
            css={[Function]}
          >
            Phones
          </div>
          <div
            className="phone-container"
          >
            <input
              className="input"
              disabled={true}
              onChange={[Function]}
              placeholder="Phone 1"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="footer"
    >
      <div
        className="flex"
        css={
          {
            "padding": "0 12px 12px 12px",
          }
        }
      >
        <button
          className="primary"
          css={[Function]}
          disabled={true}
          onClick={[Function]}
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</div>
`;
